Add a jmh module for benchmarking

msridhar opened the pull request:
This is a first crack at creating a jmh module for benchmarking NullAway, so we can more easily detect performance regressions.  We have a single benchmark for now, Caffeine 3.0.2.  `./gradlew jmh` runs the benchmark.  This is an MVP: in the future, we should add more benchmarks, make the benchmark runner more versatile, etc.

For now, running benchmarks only works on JDK 11.  We can add support for JDK 8 in the future if we want (by adding EP javac to bootclasspaths, etc.).

lazaroclapp commented on jmh/build.gradle:
```diff
@@ -0,0 +1,43 @@
+    implementation deps.build.errorProneCoreForApi
+
+    // Source jars for our desired benchmarks
+    benchmarkSources('com.github.ben-manes.caffeine:caffeine:3.0.2:sources') {
```
👍

lazaroclapp commented on jmh/build.gradle:
```diff
@@ -0,0 +1,43 @@
+compileJava.dependsOn(extractBenchmarkSources)
+
+jmh {
+//    profilers = ["stack:lines=3"]
```
Are these comments just the defaults?

lazaroclapp commented on .github/workflows/continuous-integration.yml:
```diff
@@ -42,12 +42,12 @@ jobs:
         uses: eskatos/gradle-command-action@v1
         with:
-          arguments: :nullaway:test
+          arguments: :nullaway:test :jmh:test
```
Just to clarify, does this run the performance test with caffeine? Or just the single sample file test? (I would have expected the caffeine run to be slower than what I am seeing for the CI jobs for this PR)

msridhar replied:
It only runs the single unit test `NullawayJavacTest`.  Running the performance test with Caffeine requires running `./gradlew jmh`, and the standard `build` task does not depend on the `jmh` task.

lazaroclapp commented on jmh/src/main/java/com/uber/nullaway/jmh/NullawayJavac.java:
```diff
@@ -0,0 +1,158 @@
+                + annotatedPackages);
+  }
+
+  public Boolean testCompile() {
```
What does the return value of this mean? Javadoc for public methods?

lazaroclapp commented on jmh/src/main/java/com/uber/nullaway/jmh/NullawayJavac.java:
```diff
@@ -0,0 +1,158 @@
+    return Integer.parseInt(version);
+  }
+
+  public static void main(String[] args) throws Exception {
```
Maybe a javadoc here explaining that (assuming I understand this correctly) this is not the actual entry point to the benchmark, but that it's run by the jmh gradle plugin in such and such way...?

lazaroclapp commented on jmh/src/main/java/com/uber/nullaway/jmh/NullawayJavac.java:
```diff
@@ -0,0 +1,158 @@
+  private StandardJavaFileManager fileManager;
+  private List<String> options;
+
+  public void prepareForSimpleTest() throws IOException {
```
Javadoc

lazaroclapp commented on jmh/src/main/java/com/uber/nullaway/jmh/NullawayJavac.java:
```diff
@@ -0,0 +1,158 @@
+    finishSetup("com.uber");
+  }
+
+  public void prepare(List<String> sourceFileNames, String annotatedPackages) throws IOException {
```
Javadoc

lazaroclapp commented on jmh/src/main/java/com/uber/nullaway/jmh/NullawayJavac.java:
```diff
@@ -0,0 +1,158 @@
+   * <emph>classpath</emph> for the spawned javac instance as well. So, if a benchmark depends on
+   * some library that NullAway depends on (e.g., Guava), the dependence will be magically
+   * satisfied. Note that this could lead to problems for benchmarks that depend on a conflicting
+   * version of a library.
```
Good to know. I assume there was no easy way of configuring the `processorpath` for the spawned `javac` in Gradle or otherwise pass that info. Right? (I think this should work for the cases we care about)

msridhar replied:
So now that I think about it, we could possibly do this in a slightly cleaner fashion by sticking the processor path in a JVM property (like how we pass the path to the source files).  It'll take some thinking but might be worth it.  Let me try.

msridhar replied:
I tinkered and did find a way to run NullAway without making the project depend on it, but it  had a flaw.  Without including NullAway as a dep of the project, its classes get reloaded each time the compiler runs, which really messes up performance measurement.  So we need to do things like this.  I've updated the comments.

lazaroclapp commented on jmh/src/main/java/com/uber/nullaway/jmh/NullawayJavac.java:
```diff
@@ -0,0 +1,158 @@
+    fileManager = compiler.getStandardFileManager(diagnosticListener, null, null);
+    Path outputDir = Files.createTempDirectory("classes");
+    outputDir.toFile().deleteOnExit();
+    // TODO support passing additional benchmark dependencies as the -classpath argument
```
👍

lazaroclapp reviewed (commented):
Really cool stuff! A few comments :)

msridhar commented on jmh/src/main/java/com/uber/nullaway/jmh/NullawayJavac.java:
```diff
@@ -0,0 +1,109 @@
+package com.uber.nullaway.jmh;
```
Done

msridhar commented:
@lazaroclapp I cleaned up a lot of stuff based on your feedback.  You may want to review from scratch 🙂 Also I changed the code to show how classpath dependencies of a benchmark can be passed in.  It's not necessary for Caffeine but it's a good template for adding more benchmarks in the future.

lazaroclapp commented on jmh/build.gradle:
```diff
@@ -0,0 +1,60 @@
+    warmupIterations = 10
+
+    // a trick: to get the classpath for a benchmark, create a configuration that depends on the benchmark, and
+    // then filter out the benchmark itself
```
Clever!

lazaroclapp reviewed (approved):
LGTM!
